### PR TITLE
Don't re-delete resources that are `PendingReplacement`

### LINKED ISSUE #16288: pendingReplacement
### What happened?

❯ pulumi version
v3.117.0

Two questions about when `pendingReplacement` is set to `true`

1. If I have a resource that is `"pendingReplacement": true`, pulumi tries to delete the resource on `pulumi up`. But by definition, the resource does not exist and therefore the delete fails. 
2. `pulumi refresh` does not actually call the resource's `refresh` function when `pendingReplacement` is `true`? Is there a way to change that?

Thank you for your help,
Jesse

### Example

Here is an example of a `pulumi up` with `"pendingReplacement": true` which tries to delete the resource and fails

https://app.pulumi.com/triplewhale/currency-conversion/triple-whale-staging/updates/288

### Output of `pulumi about`

❯ pulumi about  
CLI          
Version      3.117.0
Go Version   go1.22.3
Go Compiler  gc

Plugins
KIND      NAME           VERSION
resource  clickhouse     0.0.39
resource  cloudflare     5.28.0
resource  datadog        4.28.0
resource  docker         4.5.3
resource  ec             0.8.0
resource  gcp            7.24.0
resource  google-native  0.32.0
resource  grafana        0.4.2
resource  kafka          3.7.4
resource  kubernetes     4.11.0
resource  mongodbatlas   3.15.0
language  nodejs         unknown
resource  postgresql     3.11.1
resource  rabbitmq       3.3.2
resource  random         4.16.1
resource  slack          0.4.4

Host     
OS       darwin
Version  14.4.1
Arch     arm64

This project is written in nodejs: executable='/Users/jessegrodman/.nvm/versions/node/v18.17.1/bin/node' version='v18.17.1'

Current Stack: triplewhale/currency-conversion/triple-whale-staging

TYPE                                                                URN
--- sanitized ---

Found no pending operations associated with triple-whale-staging

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/triplewhale
User           triplewhale
Organizations  triplewhale
Token type     organization: triplewhale
Token type     cloud-build

Dependencies:
NAME                VERSION
@pulumi/gcp         7.24.0
@pulumi/pulumi      3.117.0
@temporalio/client  1.10.1
@tw/pulumi          0.7.138
@tw/types           0.5.470
@types/node         20.12.11

Pulumi locates its logs in /var/folders/q5/dtfqc20n7f9429fhjpj2kfxh0000gn/T/ by default

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by lunaris**
As well as indicating that a resource's state has changes, a diff can also indicate that those changes require the _replacement_ of the resource, meaning that it must be recreated and not just updated. In this scenario, there are two possible ways to replace the resource -- by first creating another new resource before deleting the old one ("create-before-replace"), or by first deleting the old resource before creating its replacement ("delete-before-replace"). Create-before-replace is the default since generally, if possible to implement, it should result in fewer instances of "downtime", where a desired resource does not exist in the system.

Should delete-before-replace be chosen, Pulumi implements this under the hood as three steps: delete for replacement, replace, and create replacement. To track things consistently, as well as enable resumption of an interrupted operation, Pulumi writes a flag, `PendingReplacement` to the state of a deleted resource that will later be cleaned up by a completed replacement.

Should an interrupted operation be resumed, Pulumi does not currently take `PendingReplacement` into account, and always enqueues a(nother) delete operation. This is typically fine (albeit wasteful) since deletes are (should) be idempotent, but unnecessary. This commit adds @jesse-triplewhale's fix for this behaviour whereby the `PendingReplacement` flag is simply removed before the remainder of the required steps (replace, create replacement) are actioned as normal. It also extends this work with some lifecycle tests for this scenario and a few others that may arise as a result of an interrupted replacement.

Fixes #16288
Closes #16303

### CODE REVIEW

File: pkg/backend/display/progress.go [hunk 1]
```diff
@@ -1229,6 +1229,8 @@ func (display *ProgressDisplay) getStepDoneDescription(step engine.StepEventMeta
 				opText = "imported"
 			case deploy.OpImportReplacement:
 				opText = "imported replacement"
+			case deploy.OpRemovePendingReplace:
+				opText = ""
```
Comment (lunaris) [inline pkg/backend/display/progress.go:1233]: I think these two changes are OK -- it seems unimportant to me to signal a purely state-level change to the user, but happy to tweak if this is a bad assumption.

File: pkg/resource/deploy/step.go [hunk 2]
```diff
@@ -1505,6 +1505,8 @@ func Color(op display.StepOp) string {
 		return colors.SpecUpdate
 	case OpReadDiscard, OpDiscardReplaced:
 		return colors.SpecDelete
+	case OpRemovePendingReplace:
+		return colors.SpecUnimportant
```
Comment (lunaris) [inline pkg/resource/deploy/step.go:1509]: Again, this feels like the right choice to me -- it's unimportant, an implementation detail (?) -- but happy to adapt if wrong.

File: pkg/resource/deploy/step.go [hunk 3]
```diff
@@ -1559,6 +1561,8 @@ func RawPrefix(op display.StepOp) string {
 		return "= "
 	case OpImportReplacement:
 		return "=>"
+	case OpRemovePendingReplace:
+		return "~ "
```
Comment (lunaris) [inline pkg/resource/deploy/step.go:1565]: We are modifying the resource's state, so I went with `~`, but any other suggestions welcome.

File: pkg/engine/lifecycletest/pending_replace_test.go [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+
+// Tests that a delete-before-replace operation:
+//
+// * that is interrupted after the deletion (e.g. with a failed create)
+// * and then resumed, with a new program that removes the deleted resource
+//
+// will:
+//
+// * not call delete (again) on the old resource
+// * remove the PendingReplacement flag from the resource in state
```
Comment (Frassle) [inline pkg/engine/lifecycletest/pending_replace_test.go:173]: Isn't this the same thing? Like if the resource is removed from state does it matter what we do with the flag?

**Review by Frassle — APPROVED** [review]

File: pkg/engine/lifecycletest/pending_replace_test.go (hunk 4 above)
Comment (lunaris) [reply pkg/engine/lifecycletest/pending_replace_test.go:173]: Yea true. This is a detail of how the steps look. Will remove this line.

**Pull request merged**